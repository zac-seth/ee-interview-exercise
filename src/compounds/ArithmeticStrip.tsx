import { FC } from 'react';

import {
  add,
  divide,
  FunctionButton,
  HandleFunctionSelected,
  multiply,
  subtract,
  VerticalStrip,
} from 'elements';

export type ArithmeticStripProps = {
  onFunctionSelected: HandleFunctionSelected;
};

const functions = [
  { label: `&divide;`, value: divide },
  { label: `&times;`, value: multiply },
  { label: `&minus;`, value: subtract },
  { label: `+`, value: add }
];

export const ArithmeticStrip: FC<ArithmeticStripProps> = ({
  onFunctionSelected
}) => (
  <VerticalStrip length={functions.length}>
    {functions.map(({ label, value }) => (
      <FunctionButton
        label={label}
        key={`function-${label}`}
        onClick={onFunctionSelected}
        value={value}
      />
    ))}
  </VerticalStrip>
);
