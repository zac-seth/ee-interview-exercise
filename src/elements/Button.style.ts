import styled, { css, FlattenSimpleInterpolation } from 'styled-components';

import { neonText, textBody } from 'tokens';

type ButtonProps = {
  isFunction?: boolean;
  isCalculate?: boolean;
  isPlaceholder?: boolean;
};


type AppliedStyles = false | FlattenSimpleInterpolation | undefined;

interface ApplyStylesForProps {
  (props: ButtonProps): AppliedStyles;
}

type CustomStyles = {
  basic?: FlattenSimpleInterpolation;
  hover?: FlattenSimpleInterpolation;
};

interface HoverStyleApplicator {
  (flag: boolean | undefined, customStyles: CustomStyles): AppliedStyles;
}

const applyStyleForFlag: HoverStyleApplicator = (flag, { basic, hover }) =>
  flag &&
  css`
    ${!!basic && basic}

    ${!!hover &&
    css`
      &:hover {
        ${hover}
      }
    `}
  `;

const applyFunctionStyles: ApplyStylesForProps = ({ isFunction }) =>
  applyStyleForFlag(isFunction, {
    basic: css`
      background-color: #1a1a1a;
    `,
    hover: css`
      background-color: #1d1d1d;
    `,
  });

const applyCalculateStyles: ApplyStylesForProps = ({ isCalculate }) =>
  applyStyleForFlag(isCalculate, {
    basic: css`
      background-color: #1a2b1a;
    `,
    hover: css`
      background-color: #1d2e1d;
    `,
  });

const applyPlaceholderStyles: ApplyStylesForProps = ({ isPlaceholder }) =>
  applyStyleForFlag(isPlaceholder, {
    basic: css`
      color: #666;
    `,
    hover: css`
      background-color: inherit;
      cursor: default;
    `,
  });

export const Button = styled.button<ButtonProps>`
  ${textBody}
  ${neonText(`1A94D2`, `228DFF`)}

  font-size: 4rem;
  line-height: 4rem;
  text-align: center;
  padding: 1rem;
  margin: 0;
  width: 8rem;
  height: 8rem;
  background-color: inherit;
  color: #1a94d2;
  border: 0;
  cursor: pointer;

  &:hover {
    background-color: #1a1a1a;
  }

  ${applyFunctionStyles}
  ${applyCalculateStyles}
  ${applyPlaceholderStyles}
`;
