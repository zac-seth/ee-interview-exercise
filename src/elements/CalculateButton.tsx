import { ReactElement } from 'react';

import { Button } from './Button.style';
import { ValuelessButtonProps } from './shared';

export const CalculateButton = ({
  label,
  onClick
}: ValuelessButtonProps): ReactElement => (
  <Button onClick={onClick} isCalculate>
    {label}
  </Button>
);
