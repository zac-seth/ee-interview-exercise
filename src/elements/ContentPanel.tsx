import { FC } from 'react';

import { Panel, PanelVariant } from './Panel.style';

export type ContentPanelProps = {
  children: JSX.Element;
};

export const ContentPanel: FC<ContentPanelProps> = ({ children }) => (
  <Panel variant={PanelVariant.CONTENT} data-testid='content-panel'>
    {children}
  </Panel>
);
