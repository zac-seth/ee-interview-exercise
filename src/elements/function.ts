export interface NumericFunction {
  (a: number, b: number): number;
}

export interface PartiallyAppliedNumericFunction {
  (b: number): number;
}

export interface CurriedNumericFunction {
  (a: number): PartiallyAppliedNumericFunction;
}

export interface Curry {
  (f: NumericFunction): CurriedNumericFunction;
}

export const curry: Curry = (f) => (a) => (b) => f(a, b);
