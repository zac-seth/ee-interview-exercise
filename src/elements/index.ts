export { add, divide, multiply, subtract } from './arithmetic';
export { CalculateButton } from './CalculateButton';
export { ContentPanel } from './ContentPanel';
export { curry } from "./function";
export { FunctionButton } from './FunctionButton';
export { Grid } from './Grid';
export { HeroPanel } from './HeroPanel';
export { HorizontalStrip } from './HorizontalStrip';
export { LayoutDirection } from './layout-direction';
export { NumberButton } from './NumberButton';
export { PlaceholderButton } from './PlaceholderButton';
export { PrimaryNumberDisplay } from './PrimaryNumberDisplay';
export { ResetButton } from './ResetButton';
export { SecondaryNumberDisplay } from './SecondaryNumberDisplay';
export { VerticalStrip } from './VerticalStrip';

export type {
  CurriedNumericFunction,
  NumericFunction,
  PartiallyAppliedNumericFunction
} from './function';
export type {
  HandleCalculateSelected,
  HandleFunctionSelected,
  HandleResetSelected,
  HandleValueSelected
} from './handler';
