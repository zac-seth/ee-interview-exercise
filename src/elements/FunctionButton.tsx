import { ReactElement } from 'react';

import { NumericFunction } from 'elements';

import { Button } from './Button.style';
import { BasicButtonProps } from './shared';

export const FunctionButton = <TValue extends NumericFunction>({
  label,
  onClick,
  value
}: BasicButtonProps<TValue>): ReactElement => (
  <Button
    onClick={() => onClick(value)}
    isFunction
    dangerouslySetInnerHTML={{ __html: label }}
  />
);
