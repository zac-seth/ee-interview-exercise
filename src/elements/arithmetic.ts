import { NumericFunction } from './function';

export const add: NumericFunction = (a, b) => a + b;

export const subtract: NumericFunction = (a, b) => a - b;

export const multiply: NumericFunction = (a, b) => a * b;

export const divide: NumericFunction = (a, b) => {
  if (b === 0) {
    throw new Error(`You can't divide by zero, fool!`);
  }

  return a / b;
};
