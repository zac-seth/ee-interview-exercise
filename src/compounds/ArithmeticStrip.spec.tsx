import { fireEvent, render, screen } from "@testing-library/react";

import { add, divide, multiply, subtract } from "elements";

import { ArithmeticStrip } from "./ArithmeticStrip";

const handleFunctionSelected = jest.fn();

describe(`ArithmeticStrip`, () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it(`Should render the arithmetic button strip`, () => {
    render(<ArithmeticStrip onFunctionSelected={() => {}} />);

    expect(screen.getByTestId(`vertical-strip-layout`)).toMatchSnapshot();
  });

  it(`Should render the correct buttons in the strip`, async () => {
    render(<ArithmeticStrip onFunctionSelected={() => {}} />);

    expect(await screen.findByText(`÷`)).toBeDefined();
    expect(await screen.findByText(`×`)).toBeDefined();
    expect(await screen.findByText(`−`)).toBeDefined();
    expect(await screen.findByText(`+`)).toBeDefined();
  });

  it(`Should correctly pass the divide function when divide is selected`, async () => {
    render(<ArithmeticStrip onFunctionSelected={handleFunctionSelected} />);

    fireEvent.click(screen.getByText(`÷`));

    expect(handleFunctionSelected).toHaveBeenCalledWith(divide);
  });

  it(`Should correctly pass the divide function when multiply is selected`, async () => {
    render(<ArithmeticStrip onFunctionSelected={handleFunctionSelected} />);

    fireEvent.click(screen.getByText(`×`));

    expect(handleFunctionSelected).toHaveBeenCalledWith(multiply);
  });

  it(`Should correctly pass the divide function when subtract is selected`, async () => {
    render(<ArithmeticStrip onFunctionSelected={handleFunctionSelected} />);

    fireEvent.click(screen.getByText(`−`));

    expect(handleFunctionSelected).toHaveBeenCalledWith(subtract);
  });

  it(`Should correctly pass the divide function when add is selected`, async () => {
    render(<ArithmeticStrip onFunctionSelected={handleFunctionSelected} />);

    fireEvent.click(screen.getByText(`+`));

    expect(handleFunctionSelected).toHaveBeenCalledWith(add);
  });
});
