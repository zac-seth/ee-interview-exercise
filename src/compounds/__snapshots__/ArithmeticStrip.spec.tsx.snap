// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArithmeticStrip Should render the arithmetic button strip 1`] = `
.c1 {
  font-family: 'Lato',sans-serif;
  font-weight: 400;
  color: #666;
  text-shadow: 0 0 5px #1A94D2,0 0 15px #228DFF;
  font-size: 4rem;
  line-height: 4rem;
  text-align: center;
  padding: 1rem;
  margin: 0;
  width: 8rem;
  height: 8rem;
  background-color: inherit;
  color: #1a94d2;
  border: 0;
  cursor: pointer;
  background-color: #1a1a1a;
}

.c1:hover {
  background-color: #1a1a1a;
}

.c1:hover {
  background-color: #1d1d1d;
}

.c0 {
  display: grid;
  grid-gap: 0.25rem;
  margin: 0;
  padding: 0;
  grid-template-rows: repeat(4,1fr);
}

<div
  class="c0"
  data-testid="vertical-strip-layout"
>
  <button
    class="c1"
  >
    ÷
  </button>
  <button
    class="c1"
  >
    ×
  </button>
  <button
    class="c1"
  >
    −
  </button>
  <button
    class="c1"
  >
    +
  </button>
</div>
`;
